### PR TITLE
Add proposal for passing ACLs in JWT

### PR DESCRIPTION
**Pull request by iblutrifork**
Proposal for an alternative Authorizer, which reads ACLs from the JWT.
Discussion on the topic: https://github.com/strimzi/strimzi-kafka-oauth/issues/223

### CODE REVIEW

File: 068-access-policies-in-jwt.md
Comment (scholzj) [inline 068-access-policies-in-jwt.md:72]: You should link to Kafka docs as Confluent Platform is not the same as Apache Kafka.

**Review by scholzj — COMMENTED** [review]
Thanks for the proposal. I think it would be great to include the following:
* Why can't you use the `type: custom` authorization with your own authorizer plugin for this?
* How will the API look like in the `Kafka` custom resource?
* What is the performance? What are the scalability limits?
* How will the code evaluating these rights look like / work?
* Are there any similar projects using this pattern?

You should also try to fix the DCO signoff -> the instructions should be inder the _Details_ link in the PR status section.

File: 068-access-policies-in-jwt.md
Comment (mstruk) [inline 068-access-policies-in-jwt.md:37]: Wouldn't that make more sense as `"acls"` rather than `"<acls>"`?

Comment (iblutrifork) [reply 068-access-policies-in-jwt.md:37]: Yes, I originally had it `<acls>` to indicate it's a configurable parameter, but I mention this in the text, so I can change this to `acls`.

File: 068-access-policies-in-jwt.md
Comment (iblutrifork) [reply 068-access-policies-in-jwt.md:64]: In the case where we want to pass the ACLs in the JWT, the ACLs should always be accessible via `getClaimsJSON()`.
I think by passing the ACLs in the JWT always, we don't need to take into account the two different flows.

File: 068-access-policies-in-jwt.md
Comment (scholzj) [inline 068-access-policies-in-jwt.md:101]: I don't think this is true. It is from my experience quite common to consume from many topics. You have also Kafka Streams applications that consume and produce many topics if they do some complicated stateful operations. So I do not think that is a flawed architecture.

**Comment by iblutrifork** [comment]
Thanks for the comments!
I'm back working on this.

> Are there any similar projects using this pattern?

Unfortunately, I cannot link to projects that I know of, as they are not public. I could look through the public strimzi forks if that is a point that is necessary to be in the proposal.

> Why can't you use the type: custom authorization with your own authorizer plugin for this?

I am not sure if I understand this question and the next one fully, can you help me out?

> How will the API look like in the Kafka custom resource?

Do you mean I have to add something similar to:

```yaml
apiVersion: kafka.strimzi.io/v1beta2
kind: Kafka
metadata:
  name: my-cluster
  namespace: myproject
spec:
  kafka:
    authorization:
      type: custom
      authorizerClass: io.mycompany.CustomAuthorizer
```

**Comment by scholzj** [comment]
> Unfortunately, I cannot link to projects that I know of, as they are not public. I could look through the public strimzi forks if that is a point that is necessary to be in the proposal.

This is not about Strimzi forks or about any private projects. It is about other products or open-source projects that are using this kind of authorization pattern. To show that this is a generally recognized pattern used out there by others.

> I am not sure if I understand this question and the next one fully, can you help me out?

Strimzi supports users to bring and use custom authorizers using the `type: custom` in `.spec.kafka.authorization`. So you should explain why that is not an option here.

File: 068-access-policies-in-jwt.md
Comment (iblutrifork) [reply 068-access-policies-in-jwt.md]: I have only once worked with consuming data from more than 20 topics, and it was not a nice experience.
Calling it flawed was my bad, I will remove the comment.

Comment (scholzj) [reply 068-access-policies-in-jwt.md]: To be honest, I think you should maybe remove the whole subjective section on how you think Kafka should be used and instead focus on the hard data of the actual scalability and limitations. I.e. things such as:
* With how many permissions is it expected to work
* With how many permissions does it break
* What is the performance and memory impact for each

Once the limits are clearly listed, people can decide on their own of how useful they are.

For me personally, I don't think 100s of permissions would be common. But 10s of them are definitely very common.

Comment (iblutrifork) [reply 068-access-policies-in-jwt.md]: I'll rework it, leaving only the numbers.

**Comment by iblutrifork** [comment]
> Strimzi supports users to bring and use custom authorizers using the type: custom in .spec.kafka.authorization. So you should explain why that is not an option here.

I've added a couple of sentences on this.

**Comment by scholzj** [comment]
> > Strimzi supports users to bring and use custom authorizers using the type: custom in .spec.kafka.authorization. So you should explain why that is not an option here.
> 
> I've added a couple of sentences on this.

I see it mentioned on line 17. But I do not see any explanation why you cannot use it.

**Comment by iblutrifork** [comment]
> I see it mentioned on line 17. But I do not see any explanation why you cannot use it.

I have added a quote from the documentation, which repeats what the paragraph around line 17 says:

Using a custom authorizer is always an option, but is more work for users, instead of having the JWT authorizer built-in. Everyone who needs to interact with a different OAuth identity provider has to implement a custom authorizer.
This is the pain point this proposal is aiming to solve.

**Comment by scholzj** [comment]
> > I see it mentioned on line 17. But I do not see any explanation why you cannot use it.
> 
> I have added a quote from the documentation, which repeats what the paragraph around line 17 says:
> 
> Using a custom authorizer is always an option, but is more work for users, instead of having the JWT authorizer built-in. 

Ok, so I take it as that nothing technical prevents users from using it as a custom authorizer.

> Everyone who needs to interact with a different OAuth identity provider has to implement a custom authorizer. This is the pain point this proposal is aiming to solve.

The proposal suggests that this will work with all kind of OAuth providers. So why do you now suggest that everyone would suddenly need to implement their own?

I don't really understand you. You are trying to convince us how great idea this is, how useful it is for everyone, how little maintenance cost it will have, etc. But you absolutely don't want to consider maintaining this yourself and insist that adding a JAR to the container image is too _cumbersome_ for anyone to use it. That does not really add up.

**Review by tombentley — COMMENTED** [review]
@iblutrifork thanks for the proposal. I think what you're aiming for would be of benefit for people wanting to use their own existing IDP without having to resort to writing their own. 

You give an example policy of `cluster_x:topic_1:read,topic2:write`. I think it would be  good to describe in more detail the proposed syntax. There's a balance to be struck there between verbosity and clarity. Obviously it needs to cover all the resource types and ACL operations. There's some variation in what's supported for attributes and roles by the different IDPs mentioned, so it would be good to confirm that whatever is proposed is actually going to be practical with those, for example by showing how an example policy maps to the format/structure supported by each IDP.

I also didn't entirely understand how the IDP's support for roles was (or was not) intended to map to Kafka's model, which is not explicitly role-based. I think a statement and maybe an example there might go a long way.

Thanks again!

File: 068-access-policies-in-jwt.md
Comment (fvaleri) [inline 068-access-policies-in-jwt.md:59]: Are we going to support regex expressions for the topic name? I think this would help to support use cases where you have many topics with a common prefix (e.g. the app name).

File: 068-access-policies-in-jwt.md
Comment (fvaleri) [inline 068-access-policies-in-jwt.md:172]:
```suggestion
If the average topic name length is 20 characters, with 10 ACLs the overhead would be 200-300 bytes.
```

File: 068-access-policies-in-jwt.md
Comment (fvaleri) [inline 068-access-policies-in-jwt.md:175]: Are we going to log warnings in these cases?

File: 068-access-policies-in-jwt.md
Comment (fvaleri) [inline 068-access-policies-in-jwt.md:160]: Can we also add a Keycloak example? Some users may want to switch to this new authorizer, and that would make their life easier, and increase adoption.

**Review by fvaleri — COMMENTED** [review]
Hi, thanks for this interesting proposal.
Left some comments.

I also think that it would be good to provide some more details on the proposed syntax for the custom ACL field.

File: 068-access-policies-in-jwt.md
Comment (ppatierno) [inline 068-access-policies-in-jwt.md:76]: why we need two different ways? Isn't it having `acls` as an array enough?

File: 068-access-policies-in-jwt.md
Comment (ppatierno) [inline 068-access-policies-in-jwt.md:67]: is this syntax a kind of standard? My understanding is that omitting the cluster name, could mean all clusters like using a regex like `.*`. Is it right? Maybe it's related to the @fvaleri question if we are going to support regex and what's the syntax for this acls.

File: 068-access-policies-in-jwt.md
Comment (iblutrifork) [reply 068-access-policies-in-jwt.md:59]: That is an interesting idea. We are currently using an implementation that is a subset of glob patterns - handling wildcard matches to allow matching a prefix or a suffix. Considering that we want to compress the ACLs as much as possible, regex could be a good fit, but my concern is that it might overcomplicate it. I cannot think of another situation apart from common prefix/suffix where you'd need the power of regex.
One more concern I have with regex is that it would be hard to debug. Glob patterns are a bit more expressive and have more consistent performance impact.

File: 068-access-policies-in-jwt.md
Comment (iblutrifork) [reply 068-access-policies-in-jwt.md:160]: That makes sense to me

File: 068-access-policies-in-jwt.md
Comment (iblutrifork) [reply 068-access-policies-in-jwt.md:76]: I propose two ways in order to cover as many identity providers as possible. While passing `acls` as an array should be preferable, some identity providers support only passing strings, from what I could tell. Looking at [Google Firebase](https://firebase.google.com/docs/auth/admin/custom-claims), I cannot tell if they support passing arrays or just bools/strings/numbers. [OneLogin](https://developers.onelogin.com/api-docs/2/api-authorization/add-claim) is similar - they definitely support strings, but I cannot find information on supporting arrays.

File: 068-access-policies-in-jwt.md
Comment (fvaleri) [reply 068-access-policies-in-jwt.md:59]: I agree on the points you made in favor of glob patterns for name matching. Can you add them to the proposal with some examples? Also, I would add regex to rejected alternatives with the motivation.

File: 068-access-policies-in-jwt.md
Comment (fvaleri) [reply 068-access-policies-in-jwt.md:175]: Right, documentation is the best place to raise these warnings.

File: 068-access-policies-in-jwt.md
Comment (mstruk) [reply 068-access-policies-in-jwt.md:67]: It should probably be something like: `CLUSTER_NAME:RESOURCE_TYPE:RESOURCE_SPEC:PERMITTED_ACTIONS`

Specifically:
* `my-cluster:Topic:topic_1:Read,Describe,DescribeConfigs`
ACL specifically for Kafka cluster named `my-cluster`, grants Read, Describe and DescribeConfigs permissions to topic called `topic_1`

* `:Group:prod_consumers_*:Describe,Delete,Read`
ACL grants Describe,Delete,Read to consumer groups that start with `prod_consumers_` on any Kafka cluster

Comment (iblutrifork) [reply 068-access-policies-in-jwt.md:67]: I've updated the proposal with more details and examples. This should cover most of the other questions on the ACL syntax in previous threads as well.

I'm using mostly the same format as @mstruk proposes.
I have opted to have the `RESOURCE_TYPE` and `PERMITTED_ACTIONS` be all lowercase.
I'd also change the separator for the `PERMITTED_ACTIONS` to `|` and keep the `,` for separating different ACLs when they are passed in a single string.
However, I have no strong opinion about the exact format.

File: 068-access-policies-in-jwt.md
Comment (iblutrifork) [reply 068-access-policies-in-jwt.md:59]: done

File: 068-access-policies-in-jwt.md
Comment (fvaleri) [inline 068-access-policies-in-jwt.md:90]: Looks like we miss some operations here.
https://kafka.apache.org/documentation/#operations_in_kafka

File: 068-access-policies-in-jwt.md
Comment (fvaleri) [inline 068-access-policies-in-jwt.md:108]: Why did you change from `,` to `|` for permitted actions? Now the symbol reads "or", but the description says "and", which is kind of weird.

File: 068-access-policies-in-jwt.md
Comment (fvaleri) [inline 068-access-policies-in-jwt.md:89]: Looks like we miss some resources here.
https://kafka.apache.org/documentation/#resources_in_kafka

File: 068-access-policies-in-jwt.md
Comment (iblutrifork) [inline 068-access-policies-in-jwt.md:81]: @tombentley Adding a section to cover how roles are mapped to kafka ACLs. I am using roles as these are supported by all IDPs. If an IDP supports more advanced mechanisms of authorization or JWT manipulation, then it could also be used.

File: 068-access-policies-in-jwt.md
Comment (iblutrifork) [reply 068-access-policies-in-jwt.md:113]: Thanks for noticing, added the missing ones.

File: 068-access-policies-in-jwt.md
Comment (mstruk) [reply 068-access-policies-in-jwt.md:81]: I think the idea is to piggyback on the roles by naming them using this custom ACL syntax. However, that may look quite hackish as the UI used to configure that may offer a drop down list of all the different ACLs and it may be quite confusing.

File: 068-access-policies-in-jwt.md
Comment (iblutrifork) [reply 068-access-policies-in-jwt.md:131]: My intuition for `|` is the same as when talking about read/write/execute file permissions in Unix, where permissions are represented as 3 bits, so if you want to enable read and write you turn on the read and write permission, which is done by OR-ing them. `100 | 010 = 110`
Again, I have no strong preference, this is the first thing that came to my mind. I would prefer not to use a `,` for AND and instead use it to separate the different ACLs. I think the `&` sign is also a bad idea as it's not as easy to read alongside words.

Comment (iblutrifork) [reply 068-access-policies-in-jwt.md:131]: I like the `+`

File: 068-access-policies-in-jwt.md
Comment (iblutrifork) [reply 068-access-policies-in-jwt.md:112]: I'm adding the other resources to the proposal. I do not fully understand if they are all applicable, but I can clear that up later.

File: 068-access-policies-in-jwt.md
Comment (iblutrifork) [reply 068-access-policies-in-jwt.md:81]: I can see why you would see this as hack-ish. Some IDPs support other ways of enriching the JWT, but they are specific to the IDP.  I chose to give the example with roles as it is what most people would know and what would work under any IDP (albeit with the drawbacks that you mention). For the same reason, I'd give the same example in the documentation.
I have used fine-grained roles like this before and can say it works fine for people.
For the UI concerns, I can talk to a team member who has more experience with this and come back to you.

File: 068-access-policies-in-jwt.md
Comment (fvaleri) [inline 068-access-policies-in-jwt.md:75]:
```suggestion
  "acls": "cluster_x:topic_1:read,:topic2:write"
```

File: 068-access-policies-in-jwt.md
Comment (fvaleri) [inline 068-access-policies-in-jwt.md:115]: I would use the same suffix for these two fields. My preference is on `_NAME`.

File: 068-access-policies-in-jwt.md
Comment (fvaleri) [inline 068-access-policies-in-jwt.md:140]: Please use a tag instead of main branch for links.

File: 068-access-policies-in-jwt.md
Comment (fvaleri) [inline 068-access-policies-in-jwt.md:214]: Broken link. Please use a tag instead of main branch.

**Review by fvaleri — APPROVED** [review]
LGTM. Thanks.

**Comment by scholzj** [comment]
I'm -0.9 on this proposal. I'm not going to -1 this proposal, so I'm not going to block this if other maintainers want to proceed with this. But I think this should be rejected for the following reasons:
* The OAuth authentication and authorization proved to be very expensive in terms of maintenance - especially compared to how many users it seems to have. This seems to be because of the high cost of maintenance of the tests as well of the Oauth libraries (updating versions, releases etc.). But also because of the bad standardization in the OAuth space where everyone seems to implement things differently and requires more and more options to handle their Oauth setup.
* The proposal does not convince me that this will be the thing used by all OAuth users out of the box as it is proposed here - without each of them requiring a special customization to make it work for them. And the proposal was not able to provide any examples where this is used in similar projects to Strimzi.
* There also seems to be a clear alternative path. The authors of the proposal can create the authorizer in their own repository and keep maintaining it. Strimzi users can then easily add their authorizer to the container images and use it as the `type: custom` authorizer. While this might not have the same level of comfort, there is nothing complicated with that - especially not if this is really the thing everyone wants to use as presented in the proposal. If needed, we can also consider adding some _directory of plugins_ to our website or docs to make it easier to advertise plugins like this to the Strimzi users. And if this proves to be successful and used by many users, this discussion can be re-opened. This seems to be also much better for the Strimzi users than accepting it and then again removing it after a few releases if it proves that the concerns are valid.
* I do not think we have enough resources to take on and maintain things like these. Especially when there is a clear path how it can live outside of Strimzi as a plugin.

File: 068-access-policies-in-jwt.md
Comment (ppatierno) [inline 068-access-policies-in-jwt.md:132]: tbh I think the last two example are not readable. I think the empty defaults (for cluster and resource) or using `topic` if no resource type is specified don't work really well imho. I would use `*` or making the resource type mandatory.

File: 068-access-policies-in-jwt.md
Comment (ppatierno) [inline 068-access-policies-in-jwt.md:114]: shortened versions are really complicated to make any sense when reading a rule imho.
Also you say using `+` as separated while in one of previous comments you were suggesting the `|`.

**Review by ppatierno — COMMENTED** [review]
I left a couple of more comments but my feeling is the same as Jakub. Why can't we just allow users to have their own plugins and using `custom` type?
What I see here is a lot of effort to maintain the code working for different IDPs but mostly testing them. Are you going to write tests for all of them as well? I am not sure it would scale well.

File: 068-access-policies-in-jwt.md
Comment (fvaleri) [reply 068-access-policies-in-jwt.md:114]: I suggested to use +, as | usually is interpreted as "or", but here actually means "and".

**Comment by fvaleri** [comment]
IMO this feature has the potential to simplify the IdP configuration (mapping between users/roles and Kafka ACLs), which would allow more users to benefit from it. Maybe we need more evidence that maintenance wouldn't be a burden, and that this approach really works fine with the most popular IdPs.

File: 068-access-policies-in-jwt.md
Comment (iblutrifork) [reply 068-access-policies-in-jwt.md:115]: I'm ok with this change. These names wouldn't be used anywhere except maybe in code and in the documentation.

File: 068-access-policies-in-jwt.md
Comment (iblutrifork) [reply 068-access-policies-in-jwt.md:132]: I somewhat agree. I think best practice would be to be explicit about all the fields. The only reasonable time to use such shorthand syntax would be when developing locally or debugging.

File: 068-access-policies-in-jwt.md
Comment (iblutrifork) [reply 068-access-policies-in-jwt.md:114]: I decided to propose the shorthand variants only to allow users with very complex and lengthy ACL lists to be able to use this feature with as little overhead as possible. I would personally advise against using them.

**Comment by iblutrifork** [comment]
Thank you for your feedback on this proposal.
Our conversation has been really helpful, and I've learned a lot from it.

I will leave a short summary to help others quickly gauge the proposal's goals.

We use a custom authorizer with Strimzi, which reads ACLs from the JWT.
We created this solution to avoid being bound to a specific IDP, such as Keycloak.
In our IDP we enrich the JWT with a list of ACLs, which the authorizer uses to authorize requests.

We thought this solution could be useful for others and thought the best way to share it would be to contribute it to the Strimzi project.

I believe we should put this proposal on hold until there is more interest from the community for something similar.

**Comment by scholzj** [comment]
Discussed on the community call on 18.4.: This can be closed. If needed, it can be reopened later to get back to it.

**Comment by yvessavoy** [comment]
Hey @iblutrifork 

We have a very similar situation in our company and before I start implementing the same custom authorizer, I wanted to ask if your code is available anywhere to take a look at?

Thanks & cheers!

**Comment by cthtrifork** [comment]
@yvessavoy  we have made a much simpler and less generic version, which lead to this proposal. We do keep it updated.
It is available here:
https://github.com/trifork/cheetah-infrastructure-utils-kafka

Do consider forking it and keeping it public, so we can inspire each other.

**Pull request closed without merging**
